arangod/Aql/Query.cpp

```diff
@@ -1353,37 +1354,70 @@ aql::ExecutionState Query::cleanupTrxAndEngines(ErrorCode errorCode) {
+      // looping.
+      _shutdownState.store(ShutdownState::None, std::memory_order_relaxed);
+    });
     // simon: no need to wait here, a commit may never issue a commit
     // request to DBServers (done by AQL layer or RestTransactionHandler)
     futures::Future<Result> commitResult = _trx->commitAsync();
     TRI_ASSERT(commitResult.isReady());
     if (commitResult.get().fail()) {
       THROW_ARANGO_EXCEPTION(std::move(commitResult).get());
     }
+    TRI_IF_FAILURE("Query::finalize_before_done") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+    // We succeeded with commit. Let us cancel the guard
+    // The state of "in progress" is now correct if we exist the method.
+    guard.cancel();
   }
 
   LOG_TOPIC("7ef18", DEBUG, Logger::QUERIES)
       << elapsedSince(_startTime)
       << " Query::finalize: before cleanupPlanAndEngine"
       << " this: " << (uintptr_t)this;
 
+
   if (_serverQueryIds.empty()) {
     _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
     return ExecutionState::DONE;
   }
   
   TRI_ASSERT(ServerState::instance()->isCoordinator());
   TRI_ASSERT(_sharedState);
-  
-  ::finishDBServerParts(*this, errorCode).thenValue([ss = _sharedState, this](Result r) {
-    LOG_TOPIC_IF("fd31e", INFO, Logger::QUERIES, r.fail() && r.isNot(TRI_ERROR_HTTP_NOT_FOUND))
-      << "received error from DBServer on query finalization: " << r.errorNumber() << ", '" << r.errorMessage() << "'";
-    _sharedState->executeAndWakeup([&] {
-      _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
-      return true;
+  try {
+    TRI_IF_FAILURE("Query::finalize_error_on_finish_db_servers") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+    }
+    ::finishDBServerParts(*this, errorCode).thenValue([ss = _sharedState, this](Result r) {
+      LOG_TOPIC_IF("fd31e", INFO, Logger::QUERIES, r.fail() && r.isNot(TRI_ERROR_HTTP_NOT_FOUND))
+        << "received error from DBServer on query finalization: " << r.errorNumber() << ", '" << r.errorMessage() << "'";
+      _sharedState->executeAndWakeup([&] {
+        _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
+        return true;
+      });
     });
-  });
-  
-  return ExecutionState::WAITING;
+    return ExecutionState::WAITING;
+  } catch (...) {
+    // In case of any error that happend in sending out the requests
+    // we simply reset to done, we tried to cleanup the engines.
+    // we only get here if something in the network stack is out of order.
+    // so there is no need to retry on cleaning up the engines, caller can continue
+    // Also note: If an error in cleanup happens the query was completed already,
+    // so this error does not need to be reported to client.
+    _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
+    // TODO: For Max: should the Level be WARN or better INFO? This will not be sever as we clean up the state.
```

Comment:
I think WARN is appropriate as to our policy: WARN is there to improve the working and performance of the database by getting rid of them. This sounds exactly like such a place, if this happens at a customer, getting rid of these warnings will improve the working and performance of the database, since for example locks can be released more quickly. INFO is too weak for this purpose.

Reply:
:+1: